File: frontend/src/lib/mathTokenizer.ts
/** 从 Markdown 文本中识别 LaTeX 公式分隔符（跳过代码块/行内代码）。 */

import { normalizeInlineLatexWhitespace } from "@/lib/latexNormalize";

export type MathSegment =
  | { kind: "text"; value: string }
  | { kind: "inline"; latex: string }
  | { kind: "block"; latex: string };

type Zone = "text" | "fenced" | "inline_code";

function isEscaped(src: string, index: number): boolean {
  let bs = 0;
  let i = index - 1;
  while (i >= 0 && src[i] === "\\") {
    bs += 1;
    i -= 1;
  }
  return bs % 2 === 1;
}

function readUntil(src: string, start: number, endSeq: string): { end: number; found: boolean } {
  let i = start;
  while (i < src.length) {
    if (src.startsWith(endSeq, i) && !isEscaped(src, i)) {
      return { end: i + endSeq.length, found: true };
    }
    i += 1;
  }
  return { end: src.length, found: false };
}

function readSingleLineInlineDollar(src: string, start: number): { latex: string; end: number } | null {
  if (src[start] !== "$" || src[start + 1] === "$") return null;
  let i = start + 1;
  while (i < src.length) {
    const ch = src[i];
    if (ch === "$" && !isEscaped(src, i)) {
      const latex = normalizeInlineLatexWhitespace(src.slice(start + 1, i).trim());
      if (!latex) return null;
      return { latex, end: i + 1 };
    }
    i += 1;
  }
  return null;
}

function readParenInline(src: string, start: number): { latex: string; end: number } | null {
  if (!src.startsWith("\\(", start)) return null;
  const { end, found } = readUntil(src, start + 2, "\\)");
  if (!found) return null;
  const latex = normalizeInlineLatexWhitespace(src.slice(start + 2, end - 2).trim());
  if (!latex) return null;
  return { latex, end };
}

function readBracketBlock(src: string, start: number): { latex: string; end: number } | null {
  if (!src.startsWith("\\[", start)) return null;
  const { end, found } = readUntil(src, start + 2, "\\]");
  if (!found) return null;
  const latex = src.slice(start + 2, end - 2).trim();
  if (!latex) return null;
  return { latex, end };
}

function readDoubleDollarBlock(src: string, start: number): { latex: string; end: number } | null {
  if (!src.startsWith("$$", start)) return null;
  const { end, found } = readUntil(src, start + 2, "$$");
  if (!found) return null;
  const latex = src.slice(start + 2, end - 2).trim();
  if (!latex) return null;
  return { latex, end };
}

/** 对单段普通文本做行内公式切分。 */
export function splitInlineMath(text: string): MathSegment[] {
  const out: MathSegment[] = [];
  let i = 0;
  let buf = "";
  const flush = () => {
    if (buf) {
      out.push({ kind: "text", value: buf });
      buf = "";
    }
  };
  while (i < text.length) {
    const paren = readParenInline(text, i);
    if (paren) {
      flush();
      out.push({ kind: "inline", latex: paren.latex });
      i = paren.end;
      continue;
    }
    const dollar = readSingleLineInlineDollar(text, i);
    if (dollar) {
      flush();
      out.push({ kind: "inline", latex: dollar.latex });
      i = dollar.end;
      continue;
    }
    buf += text[i];
    i += 1;
  }
  flush();
  return out.length ? out : [{ kind: "text", value: text }];
}

/** 将 Markdown 切分为文本块、行内公式与独立公式块。 */
export function tokenizeMathInMarkdown(markdown: string): MathSegment[] {
  const src = (markdown || "").replace(/\r\n/g, "\n");
  const out: MathSegment[] = [];
  let zone: Zone = "text";
  let fence: string | null = null;
  let i = 0;
  let buf = "";

  const flushText = () => {
    if (!buf) return;
    // 行内 $...$ 留给段落解析；此处只切 block 级公式。
    out.push({ kind: "text", value: buf });
    buf = "";
  };

  while (i < src.length) {
    if (zone === "fenced") {
      const lineEnd = src.indexOf("\n", i);
      const line = lineEnd === -1 ? src.slice(i) : src.slice(i, lineEnd);
      const trimmed = line.trim();
      if (trimmed === fence) {
        out.push({ kind: "text", value: buf + line + (lineEnd === -1 ? "" : "\n") });
        buf = "";
        zone = "text";
        fence = null;
        i = lineEnd === -1 ? src.length : lineEnd + 1;
        continue;
      }
      buf += line + (lineEnd === -1 ? "" : "\n");
      i = lineEnd === -1 ? src.length : lineEnd + 1;
      continue;
    }

    if (zone === "inline_code") {
      if (src[i] === "`" && !isEscaped(src, i)) {
        buf += "`";
        i += 1;
        zone = "text";
        continue;
      }
      buf += src[i];
      i += 1;
      continue;
    }

    if (src.startsWith("```", i)) {
      flushText();
      fence = "```";
      zone = "fenced";
      buf = "```";
      i += 3;
      continue;
    }

    if (src[i] === "`" && !isEscaped(src, i)) {
      flushText();
      zone = "inline_code";
      buf = "`";
      i += 1;
      continue;
    }

    const bracket = readBracketBlock(src, i);
    if (bracket) {
      flushText();
      out.push({ kind: "block", latex: bracket.latex });
      i = bracket.end;
      continue;
    }

    const dblock = readDoubleDollarBlock(src, i);
    if (dblock) {
      flushText();
      out.push({ kind: "block", latex: dblock.latex });
      i = dblock.end;
      continue;
    }

    buf += src[i];
    i += 1;
  }

  if (zone === "fenced") {
    out.push({ kind: "text", value: buf });
  } else {
    flushText();
  }
  return out.length ? out : [{ kind: "text", value: "" }];
}

export function escapeHtmlAttr(value: string): string {
  return value
    .replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

/** 生成仅含 data-latex 的占位 HTML（不含 KaTeX DOM，供 generateJSON 解析）。 */
export function mathSegmentsToPlaceholderHtml(segments: MathSegment[]): string {
  return segments
    .map((seg) => {
      if (seg.kind === "text") return seg.value;
      const escaped = escapeHtmlAttr(seg.latex);
      if (seg.kind === "block") {
        return `<div data-type="math-block" data-latex="${escaped}"></div>`;
      }
      return `<span data-type="math-inline" data-latex="${escaped}"></span>`;
    })
    .join("");
}

File: backend/app/services/math_tokenizer.py
"""Markdown 公式分隔符识别（与前端 mathTokenizer 对齐）。"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal


@dataclass
class MathSegment:
    kind: Literal["text", "inline", "block"]
    value: str = ""
    latex: str = ""


def _is_escaped(src: str, index: int) -> bool:
    bs = 0
    i = index - 1
    while i >= 0 and src[i] == "\\":
        bs += 1
        i -= 1
    return bs % 2 == 1


def _read_until(src: str, start: int, end_seq: str) -> tuple[int, bool]:
    i = start
    while i < len(src):
        if src.startswith(end_seq, i) and not _is_escaped(src, i):
            return i + len(end_seq), True
        i += 1
    return len(src), False


def _normalize_inline_latex(latex: str) -> str:
    """Remove hard line breaks inside inline math (e.g. ``$N\\n(i)$`` from LLM/JSON wrapping)."""
    return re.sub(r"\s*\n+\s*", "", latex or "")


def split_inline_math(text: str) -> list[MathSegment]:
    out: list[MathSegment] = []
    buf = ""
    i = 0

    def flush() -> None:
        nonlocal buf
        if buf:
            out.append(MathSegment(kind="text", value=buf))
            buf = ""

    while i < len(text):
        if text.startswith("\\(", i) and not _is_escaped(text, i):
            end, found = _read_until(text, i + 2, "\\)")
            if found:
                latex = _normalize_inline_latex(text[i + 2 : end - 2].strip())
                if latex:
                    flush()
                    out.append(MathSegment(kind="inline", latex=latex))
                    i = end
                    continue
        if text[i] == "$" and not _is_escaped(text, i) and (i + 1 >= len(text) or text[i + 1] != "$"):
            j = i + 1
            matched = False
            while j < len(text):
                if text[j] == "$" and not _is_escaped(text, j):
                    latex = _normalize_inline_latex(text[i + 1 : j].strip())
                    if latex:
                        flush()
                        out.append(MathSegment(kind="inline", latex=latex))
                        i = j + 1
                        matched = True
                    break
                j += 1
            if matched:
                continue
            buf += text[i]
            i += 1
            continue
        buf += text[i]
        i += 1
    flush()
    return out or [MathSegment(kind="text", value=text)]


def tokenize_math_in_markdown(markdown: str) -> list[MathSegment]:
    src = (markdown or "").replace("\r\n", "\n")
    out: list[MathSegment] = []
    zone = "text"
    fence: str | None = None
    i = 0
    buf = ""

    def flush_text() -> None:
        nonlocal buf
        if not buf:
            return
        # 行内 $...$ 留给段落解析（_parse_paragraph_inline）；此处只切 block 级公式。
        out.append(MathSegment(kind="text", value=buf))
        buf = ""

    while i < len(src):
        if zone == "fenced":
            line_end = src.find("\n", i)
            line = src[i:] if line_end == -1 else src[i:line_end]
            if line.strip() == fence:
                out.append(MathSegment(kind="text", value=buf + line + ("" if line_end == -1 else "\n")))
                buf = ""
                zone = "text"
                fence = None
                i = len(src) if line_end == -1 else line_end + 1
                continue
            buf += line + ("" if line_end == -1 else "\n")
            i = len(src) if line_end == -1 else line_end + 1
            continue

        if zone == "inline_code":
            if src[i] == "`" and not _is_escaped(src, i):
                buf += "`"
                i += 1
                zone = "text"
                continue
            buf += src[i]
            i += 1
            continue

        if src.startswith("```", i):
            flush_text()
            fence = "```"
            zone = "fenced"
            buf = "```"
            i += 3
            continue

        if src[i] == "`" and not _is_escaped(src, i):
            flush_text()
            zone = "inline_code"
            buf = "`"
            i += 1
            continue

        if src.startswith("\\[", i) and not _is_escaped(src, i):
            end, found = _read_until(src, i + 2, "\\]")
            if found:
                latex = src[i + 2 : end - 2].strip()
                if latex:
                    flush_text()
                    out.append(MathSegment(kind="block", latex=latex))
                    i = end
                    continue

        if src.startswith("$$", i) and not _is_escaped(src, i):
            end, found = _read_until(src, i + 2, "$$")
            if found:
                latex = src[i + 2 : end - 2].strip()
                if latex:
                    flush_text()
                    out.append(MathSegment(kind="block", latex=latex))
                    i = end
                    continue

        buf += src[i]
        i += 1

    if zone == "fenced":
        out.append(MathSegment(kind="text", value=buf))
    else:
        flush_text()
    return out or [MathSegment(kind="text", value="")]
